default/python/universe_generation/monsters.py

```diff
@@ -115,62 +118,60 @@ def generate_monsters(monster_freq, systems):
         print "...", fleet_plan.name(), ": spawn rate", fleet_plan.spawn_rate(),
         print "/ spawn limit", fleet_plan.spawn_limit(),
         print "/ effective chance", basic_chance * fleet_plan.spawn_rate(),
-
-        if len(systems) < 100:
-            # Note: The WithinStarlaneJumps condition in fp.location()
-            # is the most time costly function in universe generation
-            print "/ can be spawned at", len([s for s in systems if fleet_plan.location(s)]), "systems"
-        else:
-            print  # to terminate the print line
+        fp_location_cache[fleet_plan] = set(fleet_plan.locations(systems))
+        print ("/ can be spawned at", len(fp_location_cache[fleet_plan]),
+               "of", len(systems), "systems")
         if fleet_plan.name() in nest_name_map.values():
             statistics.tracked_monsters_chance[fleet_plan.name()] = basic_chance * fleet_plan.spawn_rate()
 
     # initialize a manager for monsters that can alter the map
     # required to prevent their placement from disjoining the map
     starlane_altering_monsters = StarlaneAlteringMonsters(systems)
 
+    # collect info for tracked monster nest valid locations
+    planets = [p for s in systems for p in fo.sys_get_planets(s)]
+    tracked_nest_valid_locations = {nest: len(fo.special_locations(nest, planets))
+                                    for nest in nest_name_map}
+
     # for each system in the list that has been passed to this function, find a monster fleet that can be spawned at
     # the system and which hasn't already been added too many times, then attempt to add that monster fleet by
     # testing the spawn rate chance
+    random.shuffle(systems)
     for system in systems:
-        # collect info for tracked monster nest valid locations
-        for planet in fo.sys_get_planets(system):
-            for nest in tracked_nest_valid_locations:
-                # print "\t tracked monster check planet: %d size: %s for nest: %20s  | result: %s"
-                # % (planet, fo.planet_get_size(planet), nest, fo.special_location(nest, planet))
-                if fo.special_location(nest, planet):
-                    tracked_nest_valid_locations[nest] += 1
-
         # collect info for tracked monster valid locations
         for fp in tracked_plan_valid_locations:
-            if fp.location(system):
+            if system in fp_location_cache[fp]:
                 tracked_plan_valid_locations[fp] += 1
 
         # filter out all monster fleets whose location condition allows this system and whose counter hasn't reached 0.
-        # Note: The WithinStarlaneJumps condition in fp.location() is
-        # the most time costly function in universe generation.
         suitable_fleet_plans = [fp for fp in fleet_plans
-                                if spawn_limits[fp]
-                                and fp.location(system)
+                                if system in fp_location_cache[fp]
```

Comment:
I'm curious: You correctly identified the bug in `specials.py`, that by placing a special the location condition of a special that has been placed before could have been invalidated, so the location conditions of all already placed specials need to be re-checked, which isn't done.

The same actually applies to space monsters. So why would you *pre-calculate* their location conditions, which makes the current bugged distribution algorithm even worse, as now placement of a monster at a location that might have become invalidated by already placed monsters becomes possible? Isn't that a step in the wrong direction?

Your discovery of that bug made me realize how botched the current implementations are, and how complex a correct solution will be that actually works. IMO we need to redesign the entire algorithm from scratch. The problem goes even further than just monsters invalidating monsters and specials invalidating specials: *everything* that gets placed during universe generation potentially invalidates the location condition of *everything* already placed that has a location condition.

And even re-checking all location conditions of all things that have one after everything has been placed doesn't solve anything, because what to do if a re-checked location condition fails? There are several ways to solve the conflict, determining the optimal one will be non-trivial. And just weeding out everything that causes a conflict will probably be to aggressive, and lead to maps that have far less monsters and specials than the selected game settings are supposed to provide.

The fundamental issue is the enormous flexibility of the location conditions. We have a set of items with location conditions, where placement of each item *is affected by and affects* all items already placed. A post-placement step which tries to automatically resolve conflicts sounds awfully complicated, and I don't want to think of the possible performance hit.

I'm really scratching my head here what to do. A thorough solution will take considerable effort, and certainly won't be something we can design and implement quickly.

A temporary solution might be to actually go with your approach here and calculate the locations conditions for all monsters before any monster is placed, and all specials before any special gets placed, and make it a rule that you can't make location conditions which involve items of the same type.

So you can't check for space monsters in the location condition of space monsters, and the same with specials. And because of the order in which things are done during universe generation, you also can't check for specials in space monster location conditions (that is, you can, but it won't do anything), but you can check for space monsters in specials location conditions.

That's a set of rules content scripters need to be aware of and should be able to work with.

If we take that road (at least as a stop-gap solution for now), which I propose we do, then you need to apply the approach implemented here for monsters also to the distribution algorithm of the specials. IMO that's the best, most consistent and most performance friendly quick solution.